Lib/idlelib/idle_test/test_pyparse.py

```diff
@@ -0,0 +1,521 @@
+        # split line test.
+        eq(start(is_char_in_string=lambda index: index < 44), 44)
+
+    def test_set_lo(self):
+        code = (
+                '"""This is a module docstring"""\n'
+                'class C():\n'
+                '    def __init__(self, a,\n'
+                '                 b=True):\n'
+                '        pass\n'
+                )
+        p = self.parser
+        p.set_str(code)
+
+        # Previous character is not a newline.
+        with self.assertRaises(AssertionError):
+            p.set_lo(5)
+
+        # A value of 0 doesn't change self.str.
+        p.set_lo(0)
+        self.assertEqual(p.str, code)
+
+        # An index that is preceded by a newline.
+        p.set_lo(44)
+        self.assertEqual(p.str, code[44:])
+
+    def test_study1(self):
+        eq = self.assertEqual
+        p = self.parser
+        setstr = p.set_str
+        study = p._study1
+
+        (NONE, BACKSLASH, FIRST, NEXT, BRACKET) = range(5)
+        TestInfo = namedtuple('TestInfo', ['string', 'goodlines',
+                                           'continuation'])
+        tests = (
+            TestInfo('', [0], NONE),
+            # Docstrings.
+            TestInfo('"""This is a completed docstring."""\n', [0, 1], NONE),
+            TestInfo('"""This is a continuation docstring.\n', [0, 1], FIRST),
+            TestInfo('"""Closing quote does not match."\n', [0, 1], FIRST),
+            TestInfo("'''This is a completed docstring.'''\n", [0, 1], NONE),
+            TestInfo('"""Bracket in docstring [\n', [0, 1], FIRST),
+            TestInfo("'''This is a continuation docstring.\n", [0, 1], FIRST),
+            TestInfo("'''This is a multiline-continutation docstring.\n\n",
+                     [0, 2], NEXT),
+            # Single-quoted strings.
+            TestInfo('"This is a completed string."\n', [0, 1], NONE),
+            TestInfo('"This is a continuation string.\n', [0, 1], NONE),
+            TestInfo('"This is a continuation string.\n\n', [0, 1, 2], NONE),
+            # Comment.
+            TestInfo('# Comment\\\n', [0, 1], NONE),  # Backslash not continuation.
+            # Brackets.
+            TestInfo('("""Complete docstring in bracket"""\n', [0, 1], BRACKET),
+            TestInfo('("""Open docstring in bracket\n', [0, 1], FIRST),
+            TestInfo('a = (1 + 2) - 5 *\\\n', [0, 1], BACKSLASH),  # No bracket.
+            TestInfo('\n   def function1(self, a,\n                 b):\n',
+                     [0, 1, 3], NONE),
+            TestInfo('\n   def function1(self, a,\\\n', [0, 1, 2], BRACKET),
+            TestInfo('\n   def function1(self, a,\n', [0, 1, 2], BRACKET),
+            TestInfo('())\n', [0, 1], NONE),                    # Extra closer.
+            TestInfo(')(\n', [0, 1], BRACKET),                  # Extra closer.
+            # For the mismatched example, it doesn't look like contination.
+            TestInfo('{)(]\n', [0, 1], NONE),                   # Mismatched.
```

Comment:
From what I can tell `_study1` is all about continuation type.  Since the mismatch example isn't code that's likely to happen 'in the wild', I don't think it's likely to cause issues.  But, if someone was playing around with the editor, they might be surprised that a non-matching bracket might close a '('.

For example, in an IDLE editor,
```python
{'a': [1, 2, 3}
```
indents 1 space upon `Enter`.  I know there's not a linter to mark it as a syntax error, but I think users might expect for the next line to align with the `{` because they thought they finished the dictionary.

Reply:
See msg 312526.  Except for possible bugs, mis-indents *are* a mark of syntax error ;-)